sharedresource/v1alpha1/types_share.go

```diff
@@ -0,0 +1,55 @@
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Share allows the backing Secret or ConfigMap to be shared across namespaces. Pods can mount the shared Secret or ConfigMap by adding a CSI volume to the pod specification that uses the "csi.sharedresource.openshift.io" CSI driver and references the share in the volume attributes:
+//
+// spec:
+//  volumes:
+//  - name: shared-secret
+//    csi:
+//      driver: csi.sharedresource.openshift.io
+//      volumeAttributes:
+//        share: my-share
+//
+//For the mount to be successful, the pod's service account must be granted permission to get the named Share object with an appropriate ClusterRole and ClusterRoleBinding.
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type Share struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	Spec ShareSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ShareList contains a list of Share objects
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type ShareList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Items           []Share `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// ShareSpec defines the desired state of Share
+type ShareSpec struct {
+	// ResourceReference references the backing object for this share.
+	// +required
+	ResourceReference corev1.ObjectReference `json:"resourceReference" protobuf:"bytes,1,opt,name=resourceReference"`
```

Comment:
corev1.ObjectReference should not be used.  See the doc on the struct itself and guidance for APIs from kube.

Reply:
Is it workth going the type/union discriminator route here?

Example

```yaml
spec:
  resource:
    type: Secret # Validate type is of Secret or ConfigMap
    # Allow only oneOf secret or configMap
    secret:
      name: my-secret
      namespace: my-namespace
```

Reply:
@adambkaplan I am not seeing anything with kubebuilder that would allow for restricting to only oneOf secret or configmap.

Reply:
I might have found it, spinning up a cluster to test with.

Reply:
> Is it workth going the type/union discriminator route here?

I think that looks like a nice, readable solution.

Reply:
Plus it gives us a nice path to expand the API in the future. For example, `projected` which lets multiple Secrets/ConfigMaps be combined into a single volume.